usr/lib/soft_stdll/soft_specific.c

```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+        return CKR_ATTRIBUTE_VALUE_INVALID;
+    }
+
+    dh = DH_new();
+    if (dh == NULL) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return CKR_FUNCTION_FAILED;
+    }
+    // Create and init BIGNUM structs to stick in the DH struct
+    bn_p = BN_new();
+    bn_g = BN_new();
+    if (bn_g == NULL || bn_p == NULL) {
+        if (bn_g)
+            BN_free(bn_g);
+        if (bn_p)
+            BN_free(bn_p);
+        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+        return CKR_HOST_MEMORY;
+    }
+    // Convert from strings to BIGNUMs and stick them in the DH struct
+    BN_bin2bn((unsigned char *) prime_attr->pValue, prime_attr->ulValueLen,
+              bn_p);
+    BN_bin2bn((unsigned char *) base_attr->pValue, base_attr->ulValueLen, bn_g);
+    dh->p = bn_p;
+    dh->g = bn_g;
+
+    // Generate the DH Key
+    if (!DH_generate_key(dh)) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return CKR_FUNCTION_FAILED;
+    }
+    // Extract the public and private key components from the DH struct,
+    // and insert them in the publ_tmpl and priv_tmpl
+
+    //
+    // pub_key
+    //
+    //temp_bn = BN_new();
+    temp_bn = dh->pub_key;
+    temp_bn_len = BN_num_bytes(temp_bn);
+    temp_byte = malloc(temp_bn_len);
+    temp_bn_len = BN_bn2bin(temp_bn, temp_byte);
+    // in bytes
+    rc = build_attribute(CKA_VALUE, temp_byte, temp_bn_len, &temp_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        return CKR_FUNCTION_FAILED;
+    }
+    template_update_attribute(publ_tmpl, temp_attr);
+    free(temp_byte);
+
+    //
+    // priv_key
+    //
+    //temp_bn = BN_new();
+    temp_bn = dh->priv_key;
+    temp_bn_len = BN_num_bytes(temp_bn);
+    temp_byte = malloc(temp_bn_len);
+    temp_bn_len = BN_bn2bin(temp_bn, temp_byte);
+    // in bytes
+    rc = build_attribute(CKA_VALUE, temp_byte, temp_bn_len, &temp_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        return CKR_FUNCTION_FAILED;
+    }
+    template_update_attribute(priv_tmpl, temp_attr);
+    free(temp_byte);
+
+    // Update CKA_VALUE_BITS attribute in the private key
+    value_bits_attr =
+        (CK_ATTRIBUTE *) malloc(sizeof(CK_ATTRIBUTE) + sizeof(CK_ULONG));
+    value_bits_attr->type = CKA_VALUE_BITS;
+    value_bits_attr->ulValueLen = sizeof(CK_ULONG);
+    value_bits_attr->pValue =
+        (CK_BYTE *) value_bits_attr + sizeof(CK_ATTRIBUTE);
+    *(CK_ULONG *) value_bits_attr->pValue = 8 * temp_bn_len;
+    template_update_attribute(priv_tmpl, value_bits_attr);
+
+    // Add prime and base to the private key template
+    rc = build_attribute(CKA_PRIME,
+                         (unsigned char *) prime_attr->pValue,
+                         prime_attr->ulValueLen, &temp_attr);  // in bytes
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
         return CKR_FUNCTION_FAILED;
```

Comment:
Shouldn't we free dh, bn_p, bn_g, and value_bits_attr here?

Reply:
Done for dh, bn_p, bn_g. The template was already updated at this place to point to value_bits_attr, so its applications responsibility to free the template in case of an error.